pkg/smokescreen/smokescreen.go

```diff
@@ -258,29 +259,61 @@ func dialContext(ctx context.Context, network, addr string) (net.Conn, error) {
 	return conn, nil
 }
 
-func rejectResponse(req *http.Request, config *Config, err error) *http.Response {
-	var msg string
-	switch err.(type) {
-	case denyError:
-		msg = fmt.Sprintf(denyMsgTmpl, req.Host, err.Error())
-	default:
-		config.Log.WithFields(logrus.Fields{
+// HTTPErrorHandler allows returning a custom error response when smokescreen
+// fails to connect to the proxy target.
+func HTTPErrorHandler(w io.WriteCloser, pctx *goproxy.ProxyCtx, err error) {
+	sctx := pctx.UserData.(*smokescreenContext)
+	resp := rejectResponse(pctx, err)
+
+	if err := resp.Write(w); err != nil {
+		sctx.logger.Errorf("Failed to write HTTP error response: %s", err)
+	}
+
+	if err := w.Close(); err != nil {
+		sctx.logger.Errorf("Failed to close proxy client connection: %s", err)
+	}
+}
+
+func rejectResponse(pctx *goproxy.ProxyCtx, err error) *http.Response {
+	sctx := pctx.UserData.(*smokescreenContext)
+
+	var msg, status string
+	var code int
+
+	if e, ok := err.(net.Error); ok {
+		// net.Dial timeout
+		if e.Timeout() {
+			status = "Gateway timeout"
+			code = http.StatusGatewayTimeout
+			msg = "Timed out connecting to remote host: " + e.Error()
+		} else {
+			status = "Bad gateway"
+			code = http.StatusBadGateway
+			msg = "Failed to connect to remote host: " + e.Error()
+		}
+	} else if e, ok := err.(denyError); ok {
+		status = "Request rejected by proxy"
+		code = http.StatusProxyAuthRequired
```

Comment:
I'm not convinced 407 Proxy Authentication Required is better than 403 Forbidden. A proxy server sending 407 is supposed to also send a `Proxy-Authenticate` header with a suggestion for the client (like "try Basic auth"), and the client is supposed to be able to interpret that suggestion and retry its request if it wants to. 403 is more like "nope, don't try again."

If you want to stick with 407, I suggest including a comment justifying that choice.

Reply:
`407 Proxy Authentication Required` has been the return code for Smokescreen ACL denials since its inception. I completely agree about the 403 being a more appropriate status code, but I guarantee there are now services relying on the old status code. I'd prefer to save this for another migration. I'll also open an issue to track it.